### PR TITLE
Verify selinux level format

### LINKED ISSUE #871: validation test failure related to SELinux
Built cri from source. Able to run unit test, integration test. while running validation tests using command `make test-cri`, one test related `SELinux` module is failing on RHEL 7.3:
`[k8s.io] SELinux runtime should support selinux [It] should error on create with wrong options`.          This test failures is observed only with `Enforcing` SELinux mode & with `Permissive` mode, All tests of SELinux module are getting skipped

## Expected Behavior
test should pass 

## Current Behavior
[BeforeEach] [k8s.io] SELinux
  /tmp/cri-install-crictl.STRb/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] SELinux
  /tmp/cri-install-crictl.STRb/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/selinux_linux.go:38
[BeforeEach] runtime should support selinux
  /tmp/cri-install-crictl.STRb/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/selinux_linux.go:47
[It] should error on create with wrong options
  /tmp/cri-install-crictl.STRb/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/selinux_linux.go:77
STEP: create a container with selinux
STEP: Get image status for image: busybox:1.28
STEP: Create container.
[AfterEach] runtime should support selinux
  /tmp/cri-install-crictl.STRb/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/selinux_linux.go:51
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] SELinux
  /tmp/cri-install-crictl.STRb/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51

тв Failure [0.803 seconds]
[k8s.io] SELinux
/tmp/cri-install-crictl.STRb/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:72
  runtime should support selinux
  /tmp/cri-install-crictl.STRb/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/selinux_linux.go:43
    should error on create with wrong options [It]
    /tmp/cri-install-crictl.STRb/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/selinux_linux.go:77

    Expected an error to have occurred.  Got:
        <nil>: nil

    /tmp/cri-install-crictl.STRb/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/selinux_linux.go:106

## Steps to Reproduce 

git clone https://github.com/containerd/cri $GOPATH/src/github.com/containerd/cri
cd $GOPATH/src/github.com/containerd/cri
git checkout v1.0.4
make install.deps
make
sudo make install
make test
make test-integration
make test-cri     

## Your Environment

* $ go version
go version go1.10.1 linux/amd64

* Operating System and version:
NAME="Red Hat Enterprise Linux Server"
VERSION="7.3 (Maipo)"

* $ getenforce
Enforcing

Can anyone please help in this?

### PR DESCRIPTION
**Pull request by miaoyq**
Fixes #871
@vitt-bagal  Could you verify this pr on your environment?

Update selinux util to b6fa367

/cc @Random-Liu @mikebrow 

Signed-off-by: Yanqiang Miao <miao.yanqiang@zte.com.cn>

### CODE REVIEW

**Comment by Random-Liu** [comment]
/test pull-cri-containerd-node-e2e
--

**Comment by Random-Liu** [comment]
If this works, let's cherrypick it into all branches.

File: pkg/server/helpers.go [hunk 1]
```diff
@@ -375,6 +376,12 @@ func initSelinuxOpts(selinuxOpt *runtime.SELinuxOption) (string, string, error)
 		return "", "", nil
 	}
 
+	// make sure the format of "level" is correct.
```
[the remark this answers is not in the document]
Comment (miaoyq) [reply pkg/server/helpers.go]: @mikebrow  I'm not sure, but if `level` is "", we can not get correct values of "processLabel" and "mountLabel" from `InitLabels()`.

File: pkg/server/helpers.go [hunk 2]
```diff
@@ -375,6 +376,12 @@ func initSelinuxOpts(selinuxOpt *runtime.SELinuxOption) (string, string, error)
...
+	matched, err := regexp.MatchString(`^s\d(\:c\d{1,4}[,|.]c\d{1,4}$)+`, selinuxOpt.GetLevel())
```
[the remark this answers is not in the document]
Comment (miaoyq) [reply pkg/server/helpers.go]: @mikebrow Thanks for your guidance.

>So I think the ",|." is wrong in so far as if it's a comma you're on to the next category, if it's a period then you get the other side of the range.

I didn't quite catch what your meaning. :-P

I did not find a complete definition of the level format, so I would also like to ask if the following formats are correct:
```
s0
s0-s0
s0-s0:c1
s0:c1,c4
s0:c1.c4
s0:c1.c4,c8
s0:c1,c4.c8
s0:c1,c4,c8
s0:c1.c4.c8
```

What about  `^s\d(-s\d)??((\:c\d{1,4})([,|.]c\d{1,4})*)?$` ?

File: pkg/server/helpers.go [hunk 3]
```diff
@@ -375,6 +376,12 @@ func initSelinuxOpts(selinuxOpt *runtime.SELinuxOption) (string, string, error)
...
+	if err != nil || !matched {
+		return "", "", fmt.Errorf("the format of 'level' is not correct: %s", selinuxOpt.GetLevel())
```
[the remark this answers is not in the document]
Comment (miaoyq) [reply pkg/server/helpers.go]: OK, will.

File: pkg/server/helpers_selinux_test.go [hunk 4]
```diff
@@ -61,12 +62,25 @@ func TestInitSelinuxOpts(t *testing.T) {
 			processLabel: "user_u:user_r:user_t:s0:c1,c2",
 			mountLabels:  []string{"user_u:object_r:container_file_t:s0:c1,c2", "user_u:object_r:svirt_sandbox_file_t:s0:c1,c2"},
 		},
+		"Should return error when the format of 'level' is not correct": {
+			selinuxOpt: &runtime.SELinuxOption{
+				User:  "user_u",
+				Role:  "user_r",
+				Type:  "user_t",
+				Level: "s0,c1,c2",
```
[the remark this answers is not in the document]
Comment (miaoyq) [reply pkg/server/helpers_selinux_test.go]: @mikebrow OK, I'll create a separate function for regexp.

**Comment by vitt-bagal** [comment]
> Fixes #871
> @vitt-bagal Could you verify this pr on your environment?

@miaoyq I have verified this pr on **Rhel 7.3** & now **test** is getting **passed**. Thanks for your help.

File: pkg/server/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikebrow) [reply pkg/server/helpers.go]: maybe "s0"?

File: pkg/server/helpers.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikebrow) [reply pkg/server/helpers.go]: All but "s0:c1.c4.c8" look good. Did you mean "s0:c1,c4.c8"?

For the categories you have a category sets, an element of this set is one of a single category or a single category followed by a "." range character and a second category, the elements are separated by commas. 

So c1,c3,c5.c6,c8 is a valid category set.

Try this one (I have not tested it and my reg is rusty):
^s\d(-s\d)(\:c\d{1,4}(\.c\d{1,4})(\,c\d{1,4}(\.c\d{1,4}))*)?$

[the remark this answers is not in the document]
Comment (miaoyq) [reply pkg/server/helpers.go]: @mikebrow Thanks, I have modified the regular expression and added a more complete test case, PTAL.

File: pkg/server/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (miaoyq) [reply pkg/server/helpers.go]: @mikebrow The latest selinux util have fixed this, I will update the selinux util to master branch.

**Comment by vitt-bagal** [comment]
@miaoyq In our environment **SELinux** tests are running on **RHEL** only. As per the my observation on  travis-ci **make test-cri** logs, tests related to **SELinux/Apparmor** module test (count-3) are not running.

**Comment by miaoyq** [comment]
>@miaoyq In my environment SELinux tests are running on RHEL only. As per the my observation on travis-ci make test-cri logs, tests related to Apparmor module (No. of test-3) are not running ubuntu.

Did you mean `Apparmor` related test cases have not been running in travis-ci? @vitt-bagal

**Comment by miaoyq** [comment]
@vitt-bagal `test-cri` run `Apparmor` related test cases only if `Apparmor` module hase been enabled on OS. 
see: https://github.com/kubernetes-incubator/cri-tools/blob/release-1.11/pkg/validate/apparmor.go#L61

I think `Apparmor` module isn't enabled by default in `travis-ci` operating system.

**Comment by mikebrow** [comment]
@vitt-bagal we'll probably cherry pick it into supported branches.

File: pkg/server/helpers.go [hunk 5]
```diff
@@ -383,6 +389,18 @@ func initSelinuxOpts(selinuxOpt *runtime.SELinuxOption) (string, string, error)
...
 }
 
+func checkSelinuxLevel(level string) (bool, error) {
+	if len(level) == 0 {
+		return true, nil
+	}
+
+	matched, err := regexp.MatchString(`^s\d(-s\d)??(:c\d{1,4}((.c\d{1,4})?,c\d{1,4})*(.c\d{1,4})?(,c\d{1,4}(.c\d{1,4})?)*)?$`, level)
+	if err != nil || !matched {
+		return false, fmt.Errorf("the format of 'level' %q is not correct: %s", level, err)
```
Comment (Random-Liu) [inline pkg/server/helpers.go]: nit: use %v for `err`.

**Comment by Random-Liu** [comment]
@miaoyq Thanks a lot for fixing this!

File: pkg/server/helpers.go (hunk 5 above)
Comment (miaoyq) [reply pkg/server/helpers.go]: @Random-Liu Done.

**Comment by miaoyq** [comment]
@Random-Liu OK

**Pull request merged**